### DISCUSSION TITLE
Framer Motion Exit Animations on Route Change

### DISCUSSION THREAD

**Discussion by robertjdominguez** [remix-run/remix#3139] [Q&A] [answered]
I posted this on Twitter a little while ago and @kentcdodds asked me to move it over here for a discussion. Fair warning: this is my first time playing around with Remix, but I like it a lot so far!

This is [the repo](https://github.com/robertjdominguez/remix-first-go) from the [tweet](https://twitter.com/_RobDominguez/status/1523667932454154240).

What it boils down to is that data from a `useLoaderData()` has no problem rendering unless I change the key for the motion component in the `root.jsx` file. Here's the relevant bits from the above repo:

```jsx
// root.jsx

import {
  Links,
  LiveReload,
  Meta,
  Scripts,
  ScrollRestoration,
} from "@remix-run/react";
import { useLocation, useOutlet } from "react-router-dom";
import { AnimatePresence, motion } from "framer-motion";

export const meta = () => ({
  charset: "utf-8",
  title: "Dominguez Dev",
  viewport: "width=device-width,initial-scale=1",
});

// Tried passing this into the key and everything works as it should
function stringGen() {
  return `hello`;
}

export default function App() {
  const outlet = useOutlet();
  return (
    <html lang='en'>
      <head>
        <Meta />
        <Links />
      </head>
      <body>
        <AnimatePresence exitBeforeEnter>
          <motion.main key={useLocation().pathname}>{outlet}</motion.main>
        </AnimatePresence>
        <ScrollRestoration />
        <Scripts />
        <LiveReload />
      </body>
    </html>
  );
}

```

Then I use a `loader()` in a route like so:

```jsx
// blog.jsx

export const loader = async () => {
  const data = await fetcher(endpoint, GET_ALL_POSTS);
  const results = await data.json();
  return results;
};
```

And then `useLoaderData()` within the component itself:

```jsx
// blog.jsx

export default function Blog() {
  const { data } = useLoaderData();
  return (
    <motion.div
      key='blog'
      initial={{ opacity: 0, y: 100 }}
      animate={{ opacity: 1, y: 0 }}
      exit={{ x: -500, opacity: 0 }}
    >
      <h1>Blog</h1>
      <ul>
        {data.posts.map((post) => (
          <li key={post.id}>
            <Link to={`/posts/${post.slug}`}>{post.title}</Link>
          </li>
        ))}
      </ul>
    </motion.div>
  );
}
```

And then, if I enter a route that needs to load data using `loader()`, I get an error message like the one below:

![image](https://user-images.githubusercontent.com/24390149/167443300-7845cd23-9c2d-4f12-8b1a-eaf5aeaaaccc.png)

Logging the value of the data from the loader to the server console doesn't return `undefined` but instead the actual data. This only happens when I try to navigate from within the application; if I hit the route directly, all data renders as expected.

It just seems to be whenever this value is changed to something dynamic - or different for each route - that the above error appears.

```jsx
// root.jsx

<motion.main key={useLocation().pathname}>{outlet}</motion.main>
```

As stated: totally new to Remix, so I'm sure - as I've heard Kent and Ryan joke - I should have read the readme more carefully 🤙  Appreciate any tips!

**Comment by robertjdominguez** [comment, marked as answer]
Who forgot to install `react-router-dom`?

![image](https://i.giphy.com/media/pU7ZQcEQ8Loxq/giphy.webp)

I guess I thought it was already installed/part of Remix under the hood? Regardless, thanks to anyone who took a look 👍

**Reply by kentcdodds** [reply]
Glad you got it working!

**Reply by sciofu** [reply]
Hello! 👋 

I am not quite sure I understand why this was happening or how exactly did you fix it. Could you elaborate a bit? 🙏

**Reply by jonathanwillisdesign** [reply]
I'm getting the same exit issue, is it a 6.4 React router thing?
